File: scala/bob/src/main/scala/bob.scala
class Bob {
  def hey(msg: String) = {
    if (is_not_empty(msg)) {
      if (is_shout(msg)) {
        "Woah, chill out!"
      } else if (is_question(msg)) {
        "Sure."
      } else {
        "Whatever."
      }
    } else {
      "Fine. Be that way!"
    }
  }

  def is_shout(msg: String) = {
    msg == msg.toUpperCase && has_upper_case(msg)
  }

  def has_upper_case(msg: String) = {
    // Is there a better way to check for regex partial matches?
    "[A-Z]+".r.findFirstIn(msg).isDefined
  }

  def is_question(msg: String) = {
    msg.endsWith("?")
  }
  def is_not_empty(msg: String) = {
    "\\S+".r.findFirstIn(msg).isDefined
  }
}

File: python/bob/bob.py
import re

class Bob(object):

  def hey (self, string):
    if self.isSilence(string): 
      return 'Fine. Be that way!'
    elif self.isShout(string):
      return 'Woah, chill out!'
    elif self.isQuestion(string):
      return 'Sure.'
    else:
      return 'Whatever.'

  def isSilence(self, string):
    return not bool(re.search(r"\S+", string))

  def isQuestion(self, string):
    return string[-1] == '?'

  def isShout(self, string):
    return string.upper() == string and re.search(r"[A-Z]+", string)
